### ISSUE TITLE
JSON type detection

### ISSUE THREAD

**Issue by mjclemente** [coldbox-modules/qb#217]
This is related to #95, but I figured different enough (and possibly easy enough) to set up a separate issue for.

Currently, when I'm trying to insert JSON into a PostgreSQL `jsonb` column, an error is thrown, because the column is JSON, and qb's type detection passes the data as varchar.

To run the insert/update without error, I can manually specify the type:

```
{json_column_name: {value: serializeJSON(some_struct), cfsqltype: "CF_SQL_OTHER"}}
```

It seems like auto inference of this type might be possible. With the caveat that you'd know the implications of this better than I, it looks like an `isJson()` check could be added to the `inferSqlType()` method in `QueryUtils.cfc`, which could switch the type to `CF_SQL_OTHER` if true. I don't think there would be any negative impact on JSON data that was being stored in varchar columns, and actual JSON typing would work for inserts/updates.

**Comment by elpete** [comment]
I have a few questions:

1. Is `CF_SQL_OTHER` safe for engines that will just insert it as text?
2. `isJSON` can be kind of slow, so I'm a little worried about running it for every binding. (It's not slow overall, but this will be called a handful of times per query and potentially dozens of times a request. (Lucee, for instance, tries to convert it to JSON and [returns `false` if it throws an exception.](https://github.com/lucee/Lucee/blob/6.0/core/src/main/java/lucee/runtime/functions/conversion/IsJSON.java) You're better off doing the try catch yourself in that case. 🙁 )
3. Does `isJSON` return consistently across engines?
4. Does `isJSON` return `true` for things like `null`, `""` (empty string), simple string values, or numbers?

Right now I'm leaning towards no just because of the second point, but there could be a more performant way to check if a string is JSON. I like the idea, but it almost feels like we need some other way to mark a column as json.  Maybe you could include an example query that would insert or query json and we can brainstorm from there.

**Comment by mjclemente** [comment]
So, in the project I'm working on we're actually using your final suggestion, where we have a convention naming JSON columns, and a hook that converts them automatically.

With regard to the questions:

1. My understanding of how `CF_SQL_OTHER` works is via a Slack conversation with Ryan Guill way back. My understanding is that it just falls back to text, or rather, just tells the JDBC driver that this data is something "OTHER". I managed to track down a [few posts online](https://stackoverflow.com/questions/14122777/postgresql-uuid-equivalent-in-coldfusion) about it, but nothing definitive, and I'd need a forum for testing the behavior I guess.
2. Didn't realize that was how Lucee handled it - yeesh. I ran a few `isJson` tests before opening the issue, but I could see it slowing things down too much with multiple calls per request. What about `isStruct`, and then converting to JSON if it is?
3. A great question. As I'm looking through Lucee [issues](https://luceeserver.atlassian.net/browse/LDEV-660), this kinds blows using `isJson` out of the water. There are definitely too many edge cases where valid json is just a text string.
4. Yeah. Unfortunately. It looks like at least for numbers. 

In summary, I'm inclined to agree with your initial assessment. The only remaining idea I had was to possibly use `isArray()` and `isStruct()`, to resolve some of the `isJson` issues, by automatically serializing those. Thoughts?

Not sure if this is what you were looking for, but here's a very basic example query: https://www.db-fiddle.com/f/q4qpM7r7xxLxVc5biM3uMB/1

**Comment by mjclemente** [comment]
Closing this now, based on your initial assessment. Thanks!
